url49584c1d41_Arrem_GGS_tgz-pJ8-com_gamezgalaxy_test_console_MainJ8/src/com/gamezgalaxy/GGS/world/Level.java

```diff
@@ -0,0 +1,424 @@
+	}
+	
+	/**
+	 * Get a block in this level
+	 * @param index
+	 *            Which block to get
+	 * @return
+	 *        The block at that index
+	 */
+	public Block getTile(int index) {
+		if (index < 0) index = 0;
+		if (index >= blocks.length) index = blocks.length - 1;
+		if (blocks[index] == null)
+			return Block.getBlock((byte)0);
+		if (blocks[index].name.equals("NULL"))
+			System.out.println("" + blocks[index].getVisableBlock());
+		return blocks[index];
+	}
+	
+	/**
+	 * Get a block at the X, Y, Z coordinates
+	 * @param x
+	 *        The X coordinate
+	 * @param y
+	 *        The Y coordinate
+	 * @param z
+	 *        The Z coordinate
+	 * @return
+	 *        The block at those coordinates
+	 */
+	public Block getTile(int x, int y, int z) {
+		return getTile(PosToInt(x, y, z));
+	}
+	
+	/**
+	 * Get how big the block array is
+	 * @return
+	 *       The size of the block array
+	 */
+	public int getLength() {
+		return blocks.length;
+	}
+	
+	/**
+	 * Set a block in this world.
+	 * If the block is a physicsblock, it will be added
+	 * to the physics tick.
+	 * This method wont send out a change to the clients.
+	 * To do this, use {@link Player#GlobalBlockChange(short, short, short, Block, Level, Server)} instead
+	 * @param b
+	 *         The block to add
+	 * @param x
+	 *        The X coordinate
+	 * @param y
+	 *        The Y coordinate
+	 * @param z
+	 *        The Z coordinate
+	 * @param server
+	 *              The server this blockchange is happening in
+	 */
+	public void setTile(Block b, int x, int y, int z, Server server) {
+		setTile(b, PosToInt(x, y, z), server);
+	}
+	
+	/**
+	 * Convert coordinates to a number that will
+	 * correspond to where the coordinates are in the
+	 * block array
+     * @param x
+	 *        The X coordinate
+	 * @param y
+	 *        The Y coordinate
+	 * @param z
+	 *        The Z coordinate
+	 * @return
+	 *        The number that will correspond to where the coordinates
+	 *        are in the block array
+	 */
+	public int PosToInt(int x, int y, int z) {
+        if (x < 0) { return -1; }
+        if (x >= width) { return -1; }
+        if (y < 0) { return -1; }
+        if (y >= height) { return -1; }
+        if (z < 0) { return -1; }
+        if (z >= depth) { return -1; }
+        return x + z * width + y * width * depth;
+    }
+	
+	public int[] IntToPos(int index) {
+		int[] toreturn = new int[3];
+		toreturn[1] = (index / width / height);
+		index -= toreturn[1]*width*height;
+		toreturn[2] = (index/width);
+		index -= toreturn[2]*width;
+		toreturn[0] = index;
+		return toreturn;
+	}
+	
+	/**
+	 * Save the level
+	 * @throws IOException
+	 *                   An IOExceptoin is thrown if there is a problem writing to the file
+	 */
+	public void Save() throws IOException {
+		if (!new File("levels").exists())
+			new File("levels").mkdir();
+		FileOutputStream fos = new FileOutputStream("levels/" + name + ".ggs");
+		GZIPOutputStream gos = new GZIPOutputStream(fos);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *gos*. The resource is referenced at lines 267, 268, 269, 270. The resource is closed at lines 271, 272, 273. There are other execution paths that do not contain closure statements, for example, when ObjectOutputStream.writeLong throws an exception. Ensure *gos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).